src/server/ftpserver.rs

```diff
@@ -50,6 +50,35 @@ use std::{fmt::Debug, future::Future, net::SocketAddr, ops::Range, path::PathBuf
 /// [`Authenticator`]: auth::Authenticator
 /// [`StorageBackend`]: storage/trait.StorageBackend.html
 pub struct Server<Storage, User>
+where
+    Storage: StorageBackend<User>,
+    User: UserDetail,
+{
+    storage: Arc<dyn (Fn() -> Storage) + Send + Sync>,
+    greeting: &'static str,
+    authenticator: Arc<dyn Authenticator<User>>,
+    data_listener: Arc<dyn DataListener>,
+    presence_listener: Arc<dyn PresenceListener>,
+    passive_ports: Range<u16>,
+    passive_host: PassiveHost,
+    collect_metrics: bool,
+    ftps_mode: FtpsConfig,
+    ftps_required_control_chan: FtpsRequired,
+    ftps_required_data_chan: FtpsRequired,
+    idle_session_timeout: std::time::Duration,
+    proxy_protocol_mode: ProxyMode,
+    logger: slog::Logger,
+    site_md5: SiteMd5,
+    shutdown: Pin<Box<dyn Future<Output = options::Shutdown> + Send + Sync>>,
+    failed_logins_policy: Option<FailedLoginsPolicy>,
+    active_passive_mode: ActivePassiveMode,
+    connection_helper: Option<OsString>,
+    connection_helper_args: Vec<OsString>,
+    pasv_listener: Arc<std::sync::Mutex<Option<tokio::net::TcpListener>>>,
+}
+
+/// Used to create [`Server`]s.  
+pub struct ServerBuilder<Storage, User>
```

Comment:
I needed to move the tls configuration out of `listen`.  It needed to happen earlier.  Yet it couldn't happen in `Server::new` because it needs to happen after methods like `ftps`.  So there needed to be _some_ step after `Server::new` and before `Server::listen` that could return a `Result`.  Splitting out a `ServerBuilder` struct seemed like the cleanest way to do it.

Reply:
OK Gotcha